### PR TITLE
Config disabling of multirecord batch updates

### PR DESCRIPTION
**Pull request by rebstar6**
in batch:
- disallow updates in batch that will update/delete an existing recordset with multiple recorddata fields
- disallow updates in batch that will create new recordsets with multiple recorddata fields

all of ^ config driven

### CODE REVIEW

File: modules/api/src/test/scala/vinyldns/api/domain/batch/BatchChangeValidationsSpec.scala [hunk 1]
```diff
@@ -1493,4 +1496,145 @@ class BatchChangeValidationsSpec
 
     result(0) shouldBe valid
   }
+
```
Comment (rebstar6) [inline modules/api/src/test/scala/vinyldns/api/domain/batch/BatchChangeValidationsSpec.scala]: not func tested, it relies on config, which isnt great

File: modules/api/src/main/scala/vinyldns/api/domain/DomainValidationErrors.scala [hunk 2]
```diff
@@ -149,4 +150,20 @@ final case class MissingOwnerGroupId(recordName: String, zoneName: String)
...
+  def message: String =
+    s"""RecordSet with name $fqdn and type ${record.typ.toString} cannot be updated in a single Batch Change
+       |because it contains multiple DNS records (${record.records.length}).""".stripMargin
+      .replaceAll("\n", " ")
+}
+
+final case class NewMultiRecordError(change: AddChangeForValidation) extends DomainValidationError {
+  def message: String =
+    s"""Multi-record recordsets are not enabled for this instance of VinylDNS.
+       |There are duplicate changes with name ${change.inputChange.inputName} and type ${change.inputChange.typ}
```
Comment (pauljamescleary) [inline modules/api/src/main/scala/vinyldns/api/domain/DomainValidationErrors.scala]: "duplicate changes" here is confusing.

We are trying to say "Cannot create a new record set with multiple records for inputName and typ" I think

File: modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeValidations.scala [hunk 3]
```diff
@@ -172,16 +175,36 @@ class BatchChangeValidations(changeLimit: Int, accessValidation: AccessValidatio
...
+      change: ChangeForValidation,
+      recordSet: RecordSet): SingleValidation[Unit] =
+    if (!multiRecordEnabled & recordSet.records.length > 1) {
+      ExistingMultiRecordError(change.inputChange.inputName, recordSet).invalidNel
+    } else ().validNel
+
+  def newRecordSetIsNotMulti(
+      change: AddChangeForValidation,
+      changeGroups: ChangeForValidationMap): SingleValidation[Unit] =
+    if (!multiRecordEnabled && changeGroups
```
Comment (pauljamescleary) [inline modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeValidations.scala]: having the `changeGroups.getList.collect.length` in an `if` is aggressive and hard to follow.

I would have that separately as a `val` so it is easier to follow

File: modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeValidations.scala [hunk 4]
```diff
@@ -199,14 +222,17 @@ class BatchChangeValidations(changeLimit: Int, accessValidation: AccessValidatio
...
+      existingRecordSets.get(change.recordKey) match {
         case Some(rs) =>
           userCanUpdateRecordSet(change, auth, rs.ownerGroupId) |+|
             ownerGroupProvidedIfNeeded(
               change,
               existingRecordSets.get(change.zone.id, change.recordName, change.inputChange.typ),
-              batchOwnerGroupId)
-        case None => RecordDoesNotExist(change.inputChange.inputName).invalidNel
+              batchOwnerGroupId) |+|
+            newRecordSetIsNotMulti(change, changeGroups) |+|
```
Comment (pauljamescleary) [inline modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeValidations.scala]: kinda strange to put "multiRecordSet" checks inside `authAndOwnerGroupValidations`

Probably makes sense to create a new local val and then update the whole of validations below

File: modules/api/src/test/scala/vinyldns/api/domain/batch/BatchChangeValidationsSpec.scala [hunk 5]
```diff
@@ -1493,4 +1496,166 @@ class BatchChangeValidationsSpec
...
+  property(
+    "validateChangesWithContext: succeed update/delete to a multi record existing RecordSet if multi enabled") {
+    val existing = List(
```
Comment (pauljamescleary) [inline modules/api/src/test/scala/vinyldns/api/domain/batch/BatchChangeValidationsSpec.scala]: More of an observation here.  We seem to be doing multiple tests in one (if I follow).  Like here we are attempting an update of a multi-record and a delete of the multi-record

The test setup and assertions become difficult.  I was wondering "why are there so many existing recordsets"?

If the test scope was narrower "succeed delete to a multi-record record set if multi enabled", then the existing should just be a single multi-record record set, and then the actual test should be a single DeleteRecordSet change.  The assertion would be a single test.

With all the things going on, the test becomes hard to follow.

There is likely a lot of precedence here and 1500 lines :scream: , just something to keep in mind that unit tests should as best as possible test one thing.

File: modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeValidations.scala (hunk 4 above)
Comment (rebstar6) [reply modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeValidations.scala]: moved out anyway, i's typed really, cname already has its own non-configged duplicate check

File: modules/api/src/test/scala/vinyldns/api/domain/batch/BatchChangeValidationsSpec.scala (hunk 5 above)
Comment (rebstar6) [reply modules/api/src/test/scala/vinyldns/api/domain/batch/BatchChangeValidationsSpec.scala]: i know, its a balance. this test file is already a PITA because you cant nest the property tests (we should eventually make this a wordspec methinks.

i kinda like the grouped tests so long as it's one concept 🤷‍♀

File: modules/api/functional_test/live_tests/batch/create_batch_change_test.py [hunk 6]
```diff
@@ -2947,3 +2891,102 @@ def test_create_batch_update_record_in_shared_zone_for_unassociated_user_in_owne
         if create_rs:
             delete_rs = shared_client.delete_recordset(shared_zone['id'], create_rs['recordSet']['id'], status=202)
             shared_client.wait_until_recordset_change_status(delete_rs, 'Complete')
+
+
+@pytest.mark.skip_production
```
Comment (mitruly) [inline modules/api/functional_test/live_tests/batch/create_batch_change_test.py]: Should we run these tests in all environments if no record changes get made?

Comment (rebstar6) [reply modules/api/functional_test/live_tests/batch/create_batch_change_test.py]: nah, depending on whats done with the config value, these will fail. should be left off and they can be manually run if wanted

**Comment by britneywright** [comment]
should add to the API configuration guide: https://www.vinyldns.io/operator/config-api
and the batch change information: https://www.vinyldns.io/api/batchchange-model.html#batch-change-information-

File: modules/api/functional_test/live_tests/batch/create_batch_change_test.py (hunk 6 above)
Comment (mitruly) [reply modules/api/functional_test/live_tests/batch/create_batch_change_test.py]: Ooh, that's right.

**Comment by rebstar6** [comment]
@britneywright making a docs PR right now as followup
